### PR TITLE
feat(rce): additional signatures for NodeJS RCE (934100)

### PR DESCRIPTION
**Pull request by karelorigin**
This PR contains fixes for bypasses that were originally found and submitted by @hussein98d in `5ZLKNU33`.

### CODE REVIEW

**Comment by dune73** [comment]
### CRS Bug Bounty PR assessment

* **Rules affected** (list rules): FIXME
* **Paranoia Level addressed** (1, 2, 3, 4, full or explain): FIXME
* **FTW passes** (yes or no) : FIXME
* **Rule(s) picked for solution** (correct or not-correct or explain) : FIXME
* **Risk for false positives** (irrelevant, adequate, substantial or explain) : FIXME
* **Regular expression quality** (inspirational, decent base, needs work, adequate or explain) : FIXME
* **Documentation** (needs work, adequate or explain) : FIXME
* **Tests** (none or some or adequate) : FIXME
&nbsp;
* **Verdict** (Unusable, inspirational, usable, almost perfect or perfect) : FIXME

_This is not meant to be final. As a CRS dev, feel free to comment below and edit this form directly. As committer or observer, feel free to comment below with feedback and we will think about updating the assessment accordingly._

**Comment by karelorigin** [comment]
Hi @lifeforms, I'd love some feedback on your feedback (ha!). I was never forwarded any `process.config` or `process.mainModule` bypasses by @hussein98d. Also, I noticed some room for improvement in this rule (in general, not just applicable to this patch).  For example, the `jsDecode` transformation is not used, even though JavaScript escape sequences can be used in NodeJS, which should make it possible to completely bypass this rule using something like:

```
x=\u0072equire("url");y=x.parse("http://x.com");c\u006fnsole.log(y)
```

I could definitely fix this but I might need a couple of days since my hands are full at the moment. Adding the `jsDecode` transformation will also allow me to simplify the regexes I've added as a bonus.

**Comment by karelorigin** [comment]
@dune73, I've noticed that the tests indeed fail. @fzipi have been looking into it but are so far unable to track down the root cause. I'll definitely get back on that.

I noticed that whitespaces are also an issue with this rule. The following payload will bypass the rule in its current state:

```javascript
this. constructor
```

Do you think it would be okay for me to add the removeWhitespace transformation + rework the regular expressions to account for that change?

Would be great if you could update the assessment after that's all finished :)

**Comment by dune73** [comment]
We'll certainly update the assessment. No worries there.

And thank you for investigating the test failure. If you can't seem to find the root cause, we can try and call the cavalry.

As for the whitespace: Please confirm that `this. constructor` is actually a potential exploit, so we really ought to catch it.

The rule itself has quite a few tests, so I think it's quite safe to play around with additional transformations. Feel free to add the `t:removeWhitespace`.

**Comment by lifeforms** [comment]
I think `this['constructor']` will also work, or not? Should the regexp be updated to catch this as well?

**Comment by karelorigin** [comment]
@dune73, thank you. @fzipi told me that he managed to track down the test issue to the transformations. After the other issue is resolved, I'll be sure to add the `removeWhitespace` transformation.

**Comment by dune73** [comment]
@fzipi are you able to provide guidance how to get the tests up and running again before you leave keyboard for a week?

I'd hate to ask you to write regular expressions on your mobile phone during a holiday...

**Comment by dune73** [comment]
@theseion: @karelorigin sees a need to have regex-assembly provide evasion protection for one or two source regexes in this PR. Can you advice?

**Comment by fzipi** [comment]
> @fzipi are you able to provide guidance how to get the tests up and running again before you leave keyboard for a week?
> 
> I'd hate to ask you to write regular expressions on your mobile phone during a holiday...

Sure, let me take a look.

**Comment by fzipi** [comment]
@dune73 We discussed this in the channel and also with @karelorigin . The base64decode is breaking the jsDecode there. We need to have a different order for transformations, with jsdecode being first, or ask @lifeforms if we want to keep the base64 decode there.

**Comment by karelorigin** [comment]
This is true. The base64Decode transformation appears to cause the issue. Applying it last would technically solve the problem, but isn't ideal since it wouldn't be able to detect cases where the base64 encoded input also has JavaScript escape sequences.

**Comment by theseion** [comment]
@dune73 @karelorigin regexp-assemble currently has two modes of evasion prevention, one for Windows and one for Unix shells. I'm not sure whether that is exactly what you want (documented here: https://coreruleset.org/docs/development/regexp_assemble/#command-line-evasion-preprocessor). Here's an example: https://github.com/coreruleset/coreruleset/blob/64e5844cb2dde349eba9529504055bc646c3377f/util/regexp-assemble/data/932100.data#L94.

There's an alternative that would also work with the current facilities: using the [`template` processor](https://coreruleset.org/docs/development/regexp_assemble/#template-preprocessor). The regex from above can be simplified to
```
process(?:(?:\.|[\['"`])env)\b
```
That gives you a fixed string that you can template like this:
```
##!> template stop-evasion (?:\.|[\['"`])
process(?:{{stop-evasion}}env)\b
##!<
```

I'm open to other ideas.

**Comment by dune73** [comment]
Thank you @fzipi. We can take it from here.

**Comment by karelorigin** [comment]
Appreciate the suggestion @theseion, I'll take a look. What about the `base64Decode` transformation @dune73? Are we cool with the potential (but still rare) bypasses that might occur if we apply that one last?

**Comment by lifeforms** [comment]
I don't think the base64Decode is a must, and I have no problems if you remove it.

I remember adding it defensively, because I've seen a lot of PHP code injections where base64 is used to encode PHP source - perhaps to bypass, but also as it's a useful string format that doesn't require escaping etc. That's why when I thought of Javascript injections, the attackers might do the same and it might be a smart failsafe to do this decode. But I have not seen it in the wild, we don't have this on many other rules either, and I think it's acceptable to remove it.

File: util/regexp-assemble/data/934100.data [hunk 1]
```diff
@@ -24,9 +24,150 @@
...
+##!<
+
+
+##! "process" payloads
+##!> assemble
+process
+##!=>
+
+##! All method names
+##!> assemble
```
Comment (theseion) [inline util/regexp-assemble/data/934100.data:58]: Please indent nested processors to improve readability (will not repeat for the same issues below)

File: util/regexp-assemble/data/934100.data [hunk 2]
```diff
@@ -24,9 +24,150 @@
...
+
```
Comment (theseion) [inline util/regexp-assemble/data/934100.data:59]:
```suggestion
##! Add function names here. Currently empty.
```

File: util/regexp-assemble/data/934100.data [hunk 3]
```diff
@@ -24,9 +24,150 @@
...
+env
+##!=< prop
+##!<
+
+##! Match method calls via their usual syntax: foo.bar()
+##!> assemble
+##! NOTE: Uncomment lines below if methods are defined
+##! \.
+##!=>
+##!    ##!> assemble
```
Comment (theseion) [inline util/regexp-assemble/data/934100.data:74]: No need to use nested processors for concatenating for borth `func` and `js-call` (will not repeat for same issues below).

File: util/regexp-assemble/data/934100.data [hunk 4]
```diff
@@ -24,9 +24,150 @@
...
+    ##!<
+##!<
+
+##! Match properties via: foo["bar"]
+##!> assemble
+    ##!> assemble
+    ##!=> js-prop-start
+    ##!<
+##!=>
+    ##!> assemble
```
Comment (theseion) [inline util/regexp-assemble/data/934100.data:98]: Why do we need to add the function names here too?

File: util/regexp-assemble/data/934100.data [hunk 5]
```diff
@@ -24,9 +24,150 @@
...
+
+##! All method names
+##!> assemble
+info
+##!=< func
+##!<
+
+##! All prop names
+##!> assemble
+
```
Comment (theseion) [inline util/regexp-assemble/data/934100.data:127]:
```suggestion
##! Add property names here. Currently empty.
```

**Review by theseion — CHANGES_REQUESTED** [review]
Very nice use of regexp-assembl! ;)

**Comment by theseion** [comment]
@karelorigin any updates on this issue?

**Comment by fzipi** [comment]
@theseion We should be able to fix this one now with the new features in regexp-assembly, right?

**Comment by theseion** [comment]
Yes, but it needs the toolchain. Regexp::Assemble will generate an invalid regex.

**Comment by theseion** [comment]
I'm picking this one up again. We have switched to crs-toolchain, so I should be able to finish it.

**Comment by theseion** [comment]
I've finalized the PR. Notes:

    - rebase onto v4.0/dev
    - add missing dynamic property access
    - add missing `argv` porperty
    - assume that every method name of `process` has a twin with `sync`
      suffix
    - add missing `require` toplevel object
    - use definitions for simple expressions, instead of assemble blocks

@karelorigin please make sure that I didn't mess up anything.

**Review by karelorigin — COMMENTED** [review]
@theseion there are still some changes that I don't entirely understand. Could you explain to me why the paranoia level was upped to 2? Also, JavaScript is case-sensitive, why are we checking for lowercase values?

**Comment by theseion** [comment]
>Also, JavaScript is case-sensitive, why are we checking for lowercase values?

I've added the `i` flag (`##!+ i`) to ignore case while matching.

>Could you explain to me why the paranoia level was upped to 2?

Good catch! TBH, I didn't change that, AFAICR. @dune73 the rule should remain at PL1, correct?

**Comment by dune73** [comment]
Very good catch. Given the rule was not moved to the PL2 range in the file, this is clearly a mistake. I suggest to retag as PL1.

**Comment by theseion** [comment]
Done

**Comment by karelorigin** [comment]
> I've added the `i` flag `(##!+ i)` to ignore case while matching.

Yes, so case-insensitivity for a case-sensitive language? 😅 

I also noticed that all the whitespaces are removed to make use of the earlier added `removeWhitespace` transformation. Are we sure we want to go this route for all entries? In my opinion, it confuses things when whitespaces are not explicitly defined in the middle of a pattern. `newfunction` being one of those. We should be fine adding them where it makes sense because of the multi match.

Do all methods on [L88](https://github.com/coreruleset/coreruleset/pull/2573/files#diff-2ed755124ce6baa89780fe8d7edab80f7d2672f57d51072c2ce1ec9d166c0fb2R88) share the optional sync prefix? If not, it might make sense to split these into two (or nested) assemble blocks. That way, we are a bit closer to our 'no false positives at PL1' objective.

**Comment by theseion** [comment]
> Yes, so case-insensitivity for a case-sensitive language? 😅

Sorry, i misread that :) I didn't really think about JS being case-sensitive, but you're right of course. I'll revert.

>In my opinion, it confuses things when whitespaces are not explicitly defined in the middle of a pattern. newfunction being one of those. We should be fine adding them where it makes sense because of the multi match.

Excellent point. The reason I removed all of them was that I realized that we would have to add tons of `\s*` everywhere to properly match (especially before and after brackets and parentheses), but then discovered that the rule already removed white space. I had not realized, however, that `multiMatch` was enabled. I will revert that change as well.

>Do all methods on [L88](https://github.com/coreruleset/coreruleset/pull/2573/files#diff-2ed755124ce6baa89780fe8d7edab80f7d2672f57d51072c2ce1ec9d166c0fb2R88) share the optional sync prefix? If not, it might make sense to split these into two (or nested) assemble blocks. That way, we are a bit closer to our 'no false positives at PL1' objective.

No, probably not. I just go very annoyed with the list and figured it wouldn't matter if we matched a couple more. The obvious additions that might cause issues are (excluding things like `chmod` that are obviously things we would want to block):
- close
- exec
- exists
- link
- open
- read
- rename
- spawn
- unlink
- write

All of these existed in the list before only with the `sync` suffix. What we match:
- `.close(`
- `.close.call(`
- `["close"]`

Personally, I think we're safe and I think the simplified maintenance of that function list is worth it. @dune73, what's your opinion?

Thanks for taking such a thorough look @karelorigin!

**Comment by dune73** [comment]
Fine from my side. Feel free to merge.

**Review by theseion — APPROVED** [review]

**Pull request merged**